runtime/onert/core/src/ir/Shape.cc

```diff
@@ -73,9 +75,9 @@ void Shape::extendRank(int to_rank)
 
 uint64_t Shape::num_elements() const
 {
-  // if dimension is 0, it means unspecified and cannot calculate the total number of elements
+  // It cannot calculate the total number of elements when it has 0 or unspecified dim.
```

Comment:
In TensorFlow, having `0` as a dim is acceptable and TF can run a graph with such tensor without error.
Also, a tensor with any dim as 0 has empty buffer (not error). 
So maybe the size should be 0 instead of error?

Reply:
@hyunsik-yoon I am not sure `0` is acceptable in tensorflow lite as tensorflow. I am  just restoring checking with 0 because it had been checked and worked in that way before your PR #1822.

Reply:
Before #1822, the code was like the following.

```
uint64_t Shape::num_elements() const
{
  // All of the nodes must have non-negative dimension
  assert(std::all_of(_dimensions.begin(), _dimensions.end(),
                     [](const int32_t &v) { return (v >= 0); }));

  return std::accumulate(_dimensions.cbegin(), _dimensions.cend(), UINT64_C(1),
                         std::multiplies<uint64_t>());
}
```

Then I treated `0` as unspecified dim so changed the code in #1822. After then you mentioned that `-1` was decision.
IMHO, if any dim is `0`, the `num_dimensions` should return `0`.

Reply:
@hyunsik-yoon No. Maybe you're confused with #1547, which I didn't know. After I see #1547, I see checking unspecified would better. I will restore to older one.

Reply:
The history is quite complicated and so is our discussion. :-)
My current understanding is.. 

- if a tensor `A` has a dim == -1 : unspecified. `num_elements(A)` cannot be calculated
- if a tensor `A` has a dim == 0 : _not_ error. `num_elements(A)` is 0

Reply:
@hyunsik-yoon 

The behaviour of `num_elements()` has been changed twice by your PR #1547 and #1859.

I thought #1859 changed the behaviour of `num_elements()` mistakenly because as the title said (**"Set unknown dim (unspecified dim) to -1"**, I thought it is about introducing `UNSPECIFIED_DIM`. However, it changed `0` to `UNSPECIFIED_DIM` (= `-1`) in several places. In this PR, I tried to roll back the change. However, it seems you intended the change.

>  if a tensor A has a dim == -1 : unspecified. num_elements(A) cannot be calculated
>  if a tensor A has a dim == 0 : not error. num_elements(A) is 0

Thus, I will not change the behavior of `num_elements()` in this PR.
However, I am not sure it is correct.